### DISCUSSION TITLE
Is there any performance overhead when the request Uri has more than 2500 characters？

### DISCUSSION THREAD

**Discussion by LoremipsumSharp** [Kong/kong#8838] [Help] [answered]
We currently using Kong as API Gateway（`2.7.1`）, to route the traffic to our backend service.
About 3000 routes is maintained and most of them is `regex route`.

Everything works fine before we connected a new API to Kong last week, the CPU usage of Kong got a significant increment

The recent connected API has a special characteristic which would be invoke by a Uri with more than 3000 characters, moreover, such long Uri is contributed by many query parameters that is changed each time 

According to the  [router.lua](https://github.com/Kong/kong/blob/2.7.1/kong/router.lua) ,I found Kong would look up the cache firstly, if not ,then loop through the predefined routed to compare. Because the request Uri is changed each time, so there must be a cache miss,(if I was wrong please correct me ) ,I wonder if there is any performance issues in this context。
Before:
![](https://raw.githubusercontent.com/LoremipsumSharp/Images/master/img/before.png)
After:
![](https://raw.githubusercontent.com/LoremipsumSharp/Images/master/img/after.png)

**Comment by ms2008** [comment, marked as answer]
You've got a point.

Behind the hood, everytime kong receives a query, it tries to match it against all existing regex uris: https://github.com/Kong/kong/blob/15c427e090452396a5f52bf6848d8ad6bbe5e2d1/kong/router.lua#L1769

If the route doesn't exist, it calls `re_find` on all of them, which makes the openresty regex module automatically tries to compile and cache it in its internal lrucache. This behavior is CPU-intensive.
